### ISSUE TITLE
Provide simple API to programmatically instantiate a Tendermint node

### ISSUE THREAD

**Issue by thanethomson** [tendermint/tendermint#3690]
As per conversation in #3682, it appears as though some Tendermint users are currently importing the [`rpctest` library code](https://github.com/tendermint/tendermint/tree/master/rpc/test) directly to configure and instantiate Tendermint nodes programmatically. To me, this speaks to the need for a simpler official (perhaps somewhat opinionated) API to quickly and programmatically instantiate a Tendermint node.

We should look at providing such a simple interface as an official API (perhaps within the `node` package?), document it, and then make use of this API from the RPC tests.

**Comment by ruseinov** [comment]
@ethanfrey ^^

**Comment by ruseinov** [comment]
@thanethomson I'm working on defining such an api and once that is done - will post this here. If it gets approved - I'll be happy to implement it.

**Comment by thanethomson** [comment]
Awesome! 🎉

**Comment by thanethomson** [comment]
@ruseinov: when you get a chance to take a look at this API, my recommendation would be to start with high-level method definitions and descriptions of those methods first, without any implementation details. We can hash out implementation details in a PR, otherwise it'll end up being too much work.

One thing I'd definitely recommend is to find a way to avoid making people call `GetFreePort` at all. An easy way to configure, start and stop a Tendermint node would be a great start.

**Comment by ethanfrey** [comment]
Yes, I work with @ruseinov and was the one to pull in rpctest code into our codebase, so we could do proper end-to-end intergration tests in our go unit tests. It worked for around a year, but has been showing some issues now and then.

I really like the idea of a simple interface to provide this functionality in a public way. It looks like `tendermint/node` has been refactored in the last time (several months) to avoid the use of many globals and now defines quite a few nice helpers used in `node_test.go`.

@ruseinov I think taking a look at that code, and figuring out how to start an rpc server (no grpc or p2p needed for test cases) on top of it would be a good place to start. https://github.com/tendermint/tendermint/blob/master/node/node.go#L79-L85 looks like a pretty good API, just figure out what else is needed to make a proper test api (config builder, clean shutdown methods, etc)

**Comment by ruseinov** [comment]
Yeah, I was not sure if we want to extend this work to tendermint start/stop, so that's why I have only provided the config wrapper interface.

I'm glad we are willing to extend this further as it would make sense, I will provide some more info this week and get the ball rolling.

About GetFreePorts - I don't really see the reason for a user to call this, unless they want to somehow use this helper in their codebase for something else. To me it's just a nice way to allocate test ports for pretty much anything, but I agree that we might be better off limiting the scope of that to ConfigWrapper and making it private. 

Thanks for the input!

**Comment by ruseinov** [comment]
Alright, I'm going to duplicate some of the stuff I wrote in another issue here:

```
// this would be a signature to start a node if we want to reuse it somehow outside of configWrapper scope
func StartTendermint(c *cfg.Config, app types.Application) (node, shutdownCleanupFunc, error) {

}

// this is the signature that depends on ConfigWrapper and does it's cleanup too
func StartTendermint(c *cfg.ConfigWrapper, app types.Application) (node, shutdownCleanupFunc, error) {

}

// we have found it useful to throw some meaningful test errors when operating on tendermint // nodes/cleanup. it's just a suggestion, not a requirement though. Might be too specific to the use-case.
func StartTendermint(c *cfg.ConfigWrapper, app types.Application, t *testing.T) (node, shutdownCleanupFunc, error) {

}


type ConfigWrapper struct {
	c *cfg.Config
	rootDir string
}

// returns chainID which otherwise is private
func(cw *ConfigWrapper) ChainID() string {
}

// gets ports, assigns them to proper config fields etc
func NewConfigWrapper(testName, chainID string) (*ConfigWrapper, error) {

}

func(_ *ConfigWrapper) addr(port int) string {
}

// GetFreePorts gets n free ports while listening to all of them before they are returned
// to avoid collisions
func GetFreePorts(n int) ([]int, error) {
}


// returns underlying Config, which could be directly modified
func(cw *ConfigWrapper) Config() *cfg.Config {
}

// Cleanup makes sure nothing is left on the disk after tests have been run
func(cw *ConfigWrapper) Cleanup() {

}
```

**Comment by thanethomson** [comment]
Thanks for this! A couple of points/questions:

1. What's the difference here between `ConfigWrapper.rootDir` and `Config.BaseConfig.RootDir`?
2. Where will the port allocation take place? (It's imperative, to avoid issues like the ones you previously experienced with collisions, to minimize the time between allocating the ports and using them. The ideal would be to eliminate the need to free the listener altogether somehow)
3. We only need to know about public methods in the API definition
4. I'd recommend against including a version of `StartTendermint` that includes a `*testing.T` parameter - rather just return any errors and let those be handled during your test. The API should be usable during testing, and outside of testing, and should be as clean and simple as possible.
5. There's already a `BaseConfig.ChainID()` method (see `config/config.go`).
6. What does the `shutdownCleanupFunc()` do, when comparing that with your `ConfigWrapper.Cleanup()` method?

**Comment by usamahamid** [comment]
Is this issue still open?

@ruseinov, are you working on it? 
Just asking since it think this would be a good first issue to get started.

**Comment by ruseinov** [comment]
@usamahamid Not really, I'm working on a totally different setup now, so unfortunately don't have the time/resources to look into this any time soon.

@husio if you are interested in working on this - please feel free to take over.

**Comment by jack17529** [comment]
@ruseinov I'm sorry but there is no one named husio in the repository, correct me if I am wrong.

**Comment by ruseinov** [comment]
@jack17529 That is my colleague at IOV who worked on a similar issue within our codebase. It seems like he’s not working on this anymore either.
